src/solvers/solver_utils.hpp

```diff
@@ -135,6 +145,155 @@ struct Stencil {
+TaskStatus AddFieldsAndStore(const std::shared_ptr<MeshData<Real>> &md, Real wa = 1.0,
+                             Real wb = 1.0) {
+  return AddFieldsAndStoreInteriorSelect<a_t, b_t, out, only_fine_on_composite>(
+      md, wa, wb, false);
+}
+
+template <class var, bool only_fine_on_composite = true>
+TaskStatus SetToZero(const std::shared_ptr<MeshData<Real>> &md) {
+  int nblocks = md->NumBlocks();
+  using TE = parthenon::TopologicalElement;
+  TE te = TE::CC;
+  std::vector<bool> include_block(nblocks, true);
+  auto desc = parthenon::MakePackDescriptor<var>(md.get());
+  auto pack = desc.GetPack(md.get(), include_block, only_fine_on_composite);
+  const size_t scratch_size_in_bytes = 0;
+  const int scratch_level = 1;
+  const int ng = parthenon::Globals::nghost;
+  parthenon::par_for_outer(
+      DEFAULT_OUTER_LOOP_PATTERN, "SetFieldsToZero", DevExecSpace(),
+      scratch_size_in_bytes, scratch_level, 0, pack.GetNBlocks() - 1,
+      KOKKOS_LAMBDA(parthenon::team_mbr_t member, const int b) {
+        auto cb = GetIndexShape(pack(b, te, 0), ng);
+        const auto &coords = pack.GetCoordinates(b);
+        IndexRange ib = cb.GetBoundsI(IndexDomain::interior, te);
+        IndexRange jb = cb.GetBoundsJ(IndexDomain::interior, te);
+        IndexRange kb = cb.GetBoundsK(IndexDomain::interior, te);
+        parthenon::par_for_inner(parthenon::inner_loop_pattern_simdfor_tag, member, kb.s,
+                                 kb.e, jb.s, jb.e, ib.s, ib.e, [&](int k, int j, int i) {
+                                   const int nvars = pack.GetUpperBound(b, var()) -
+                                                     pack.GetLowerBound(b, var()) + 1;
+                                   for (int c = 0; c < nvars; ++c)
+                                     pack(b, te, var(c), k, j, i) = 0.0;
+                                 });
+      });
+  return TaskStatus::complete;
+}
+
+template <class a_t, class b_t>
+TaskStatus DotProductLocal(const std::shared_ptr<MeshData<Real>> &md,
+                           AllReduce<Real> *adotb) {
+  using TE = parthenon::TopologicalElement;
+  TE te = TE::CC;
+  IndexRange ib = md->GetBoundsI(IndexDomain::interior, te);
+  IndexRange jb = md->GetBoundsJ(IndexDomain::interior, te);
+  IndexRange kb = md->GetBoundsK(IndexDomain::interior, te);
+
+  auto desc = parthenon::MakePackDescriptor<a_t, b_t>(md.get());
+  auto pack = desc.GetPack(md.get());
+  Real gsum(0);
+  parthenon::par_reduce(
+      parthenon::loop_pattern_mdrange_tag, "DotProduct", DevExecSpace(), 0,
+      pack.GetNBlocks() - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i, Real &lsum) {
+        const int nvars = pack.GetUpperBound(b, a_t()) - pack.GetLowerBound(b, a_t()) + 1;
+        for (int c = 0; c < nvars; ++c)
+          lsum += pack(b, te, a_t(c), k, j, i) * pack(b, te, b_t(c), k, j, i);
+      },
```

Comment:
Similar to the question above: Are we leaving out vectorization here because of the explicit inner loop (rather than using hierarchical parallelism)?

Reply:
yes, see above.